### PR TITLE
Add Full Sensor Grid to Ocular plugin

### PR DESCRIPTION
**Pull request by mk-dhia**
### Description
Inside Crop overlay add a new checkbox to have the full sensor grid so we can see pixel width on the image and have a fair estimate of detail. I fund my self playing with crop sizes on image to estimate what detail level i would expect from my setup. Now you can see immediately the projected pixel size, check how many pixel is the object
in the process i made a change to the crop to add an upper limit of the sensor size 
Also i have to round the crop size to be multiple of binned pixel

I have changed naming of crop size as it was using Hsize/VSize whereas sensor/binning/fov do use X/Y so i opted for X/Y naming in my variable and i can add a commit to rename all ccdCropOverlayHSize,... to ccdCropOverlayX,... (or use HSize and VSize in my variable name, what ever you prefer)

PS: i'm not used to c++/Qt and i m willing to do the commit correctly (if you find idea worthy) do not hesitate to tell me what to fix/improve

### Screenshots (if appropriate):
(new option is show pixel grid)
image
with higer resolution sensor zooming in
image

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update

### How Has This Been Tested?
unit test ok
manual test on y laptop

**Test Configuration**:
* Operating system: <Linux Mint 19.2>
* Graphics Card: <Intel, kernel 5.3> 1920x1080

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by alex-w** [comment]
Thanks for patch!

You marked item “bug fix” - which bug you meant?

**Comment by alex-w** [comment]
You use term “full grid”, but by fact you propose the patch, which show pixel grid - why you did not used term “pixel grid” for methods and tooltip?

P.S. The code require cosmetic changes (a “stylish” changes) and one small addition to support a shortkey to toggle pixel grid.

**Comment by gzotti** [comment]
I tried to set a box size of 32x32 pixels. However, I see the box on screen labeled 32x30, and rendered as 32x10. (i.e., vertically elongated pixels.)
... 
Ah, that was a Binning setting. Maybe this should be somehow added to the box label? ("32px x 30px (Binning 1x3)")

**Comment by alex-w** [comment]
@gzotti I can add binning info into GUI bar of plugin (top right corner of the screen)

**Comment by gzotti** [comment]
I think you have more space in the right bottom corner under the grid box. But either solution is OK.

**Comment by alex-w** [comment]
> I think you have more space in the right bottom corner under the grid box. But either solution is OK.

Yes + I'll preferred to show main properties of CCD near other similar data. In any case showing a binning data is task not for this PR.

**Comment by gzotti** [comment]
I think "full grid" is just in comparison against "just a box". 's/full grid/pixel grid/g'
Ticking off "bug" is IMHO a cosmetic glitch in the PR request.

**Comment by mk-dhia** [comment]
> Thanks for patch!
> 
> You marked item “bug fix” - which bug you meant?

The crop goes beyond the sensor size if you have a sensor of 1000x1000 and you set crop of 2000x2000 the "croped" part will be larger than sensor

**Comment by mk-dhia** [comment]
> I think "full grid" is just in comparison against "just a box". 's/full grid/pixel grid/g'
> Ticking off "bug" is IMHO a cosmetic glitch in the PR request.

Ill fix that

**Comment by mk-dhia** [comment]
> You use term “full grid”, but by fact you propose the patch, which show pixel grid - why you did not used term “pixel grid” for methods and tooltip?
> 
> P.S. The code require cosmetic changes (a “stylish” changes) ...
If you put them in review i fix them or tell me what i should look for
> ..and one small addition to support a keyboard shortcut to toggle pixel grid.
I m not sure that i master  how to add that but i think i could manage it. is there a guidance on what default shortcut shall i put?

File: plugins/Oculars/src/Oculars.cpp [hunk 1]
```diff
@@ -1596,8 +1605,23 @@ void Oculars::paintCCDBounds()
 					a = transform.map(QPoint(static_cast<int>(overlayWidth*0.5f), static_cast<int>(overlayHeight*0.5f)));
 					b = transform.map(QPoint(static_cast<int>(overlayWidth*0.5f), static_cast<int>(-overlayHeight*0.5f)));
 					painter.drawLine2d(a.x(), a.y(), b.x(), b.y());
+					
+					//Tool to show full CCD grid overlay
+					if(flagShowCcdCropOverlayPixelGrid) {
```
Comment (alex-w) [inline plugins/Oculars/src/Oculars.cpp:1610]: Please move brace on new line

File: plugins/Oculars/src/Oculars.cpp [hunk 2]
```diff
@@ -1596,8 +1605,23 @@ void Oculars::paintCCDBounds()
...
+						//vertical lines
+						for (int l =1 ; l< actualCropOverlayX/ccd->binningX(); l++ ){
```
Comment (alex-w) [inline plugins/Oculars/src/Oculars.cpp:1612]: Please move brace on new line

File: plugins/Oculars/src/Oculars.cpp [hunk 3]
```diff
@@ -1596,8 +1605,23 @@ void Oculars::paintCCDBounds()
...
+							a = transform.map(QPoint(static_cast<int>(overlayWidth*0.5f- l*pixelProjectedWidth), static_cast<int>(-overlayHeight*0.5f)));
+							b = transform.map(QPoint(static_cast<int>(overlayWidth*0.5f- l*pixelProjectedWidth), static_cast<int>(overlayHeight*0.5f)));
+							painter.drawLine2d(a.x(), a.y(), b.x(), b.y());
+						}
+						//horizontal lines
+						for (int l =1 ; l< actualCropOverlayY/ccd->binningY(); l++ ){
```
Comment (alex-w) [inline plugins/Oculars/src/Oculars.cpp:1618]: Please move brace on new line

**Comment by alex-w** [comment]
Hmm... should be crop box have a size 2x if binning 2x2 is used? Looks like bug in our old code. @gzotti ?

**Comment by alex-w** [comment]
> I m not sure that i master how to add that but i think i could manage it. is there a guidance on what default shortcut shall i put?

I’ll add this part after merge.

**Comment by gzotti** [comment]
> 
> 
> Hmm... should be crop box have a size 2x if binning 2x2 is used? Looks like bug in our old code. @gzotti ?

According to the GUI label "Overlay size (px)", native pixels are OK. Or relabel to "Overlay size (binned px)" and define the box in current binning settings. For the grid, it is indeed good to show the pixel grid with current binning setting (and clarify also in the screen label). Not sure what other software titles have here, or whether there is any standard recommendation to follow?

**Comment by alex-w** [comment]
I have a strange question - why pixel grid do not use a full CCD frame?

**Comment by mk-dhia** [comment]
> I have a strange question - why pixel grid do not use a full CCD frame?

That was my first implem. however you get red rectangle when you are zoomed out (unless you use very low resolution sensor ). i found it more useful when just applied to the crop so i still see the global image

**Comment by alex-w** [comment]
OK, let’s merge it and ask users to test this feature  with feedback :)

**Comment by gzotti** [comment]
Yes, I agree using the grid only for the crop area is better. Its application is likely planetary hi-res work.

**Pull request merged**